src/Runner.Listener/MessageListener.cs

```diff
@@ -183,27 +199,52 @@ public async Task<TaskAgentMessage> GetNextMessageAsync(CancellationToken token)
+                jobStatusNotification = new TaskCompletionSource<bool>();
                 TaskAgentMessage message = null;
+                CancellationTokenSource getMessagesTokenSource = CancellationTokenSource.CreateLinkedTokenSource(token);
                 try
                 {
-                    message = await _runnerServer.GetAgentMessageAsync(_settings.PoolId,
+                    _term.WriteLine($"Getting messages with runner status as {runnerStatus}.");
+                    var getMessageTask = _runnerServer.GetAgentMessageAsync(_settings.PoolId,
                                                                 _session.SessionId,
                                                                 _lastMessageId,
-                                                                token);
+                                                                runnerStatus,
+                                                                getMessagesTokenSource.Token);
+                    
+                    var completedTask = await Task.WhenAny(getMessageTask, jobStatusNotification.Task);
 
-                    // Decrypt the message body if the session is using encryption
-                    message = DecryptMessage(message);
-
-                    if (message != null)
+                    if (completedTask == getMessageTask)
                     {
-                        _lastMessageId = message.MessageId;
-                    }
+                        if (getMessageTask.IsCanceled)
+                        {
+                            throw new OperationCanceledException();
+                        }
 
-                    if (encounteringError) //print the message once only if there was an error
+                        if (getMessageTask.IsFaulted)
+                        {
+                            throw getMessageTask.Exception.InnerException;
+                        }
+                        
+                        message = getMessageTask.Result;
+                        // Decrypt the message body if the session is using encryption
+                        message = DecryptMessage(message);
+
+                        if (message != null)
+                        {
+                            _lastMessageId = message.MessageId;
+                        }
+
+                        if (encounteringError) //print the message once only if there was an error
+                        {
+                            _term.WriteLine($"{DateTime.UtcNow:u}: Runner reconnected.");
+                            encounteringError = false;
+                            continuousError = 0;
+                        }
+                    }
+                    else 
                     {
-                        _term.WriteLine($"{DateTime.UtcNow:u}: Runner reconnected.");
-                        encounteringError = false;
-                        continuousError = 0;
+                        getMessagesTokenSource.Cancel();
```

Comment:
you might want to still `await getMessageTask;` and catch/trace all exceptions from it.